packages/component-base/src/overlay-class-mixin.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * @license
+ * Copyright (c) 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+
+/**
+ * A mixin that forwards CSS class names to the internal overlay element
+ * by setting the `overlayClass` property or `overlay-class` attribute.
+ *
+ * @polymerMixin
+ */
+export const OverlayClassMixin = (superclass) =>
+  class OverlayClassMixinClass extends superclass {
+    static get properties() {
+      return {
+        /**
+         * A space-delimited list of CSS class names to set on the overlay element.
+         * This property does not affect other CSS class names set manually via JS.
+         *
+         * @attr {string} overlay-class
+         */
+        overlayClass: {
+          type: String,
+        },
+
+        /**
+         * An overlay element on which CSS class names are set.
+         *
+         * @protected
+         */
+        _overlayElement: {
+          type: Object,
+        },
+      };
+    }
+
+    static get observers() {
+      return ['__updateOverlayClassNames(overlayClass, _overlayElement)'];
+    }
+
+    /**
+     * @param {HTMLElement} overlay
+     * @param {string[]} classInfo
+     * @private
+     */
+    __applyClassNames(overlay, classInfo) {
+      classInfo.forEach((name) => {
+        if (!this._customClasses.has(name)) {
+          overlay.classList.add(name);
+        }
+      });
+    }
+
+    /**
+     * @param {HTMLElement} overlay
+     * @param {string[]} classInfo
+     * @private
+     */
+    __clearClassNames(overlay, classInfo) {
+      if (this._previousClasses) {
+        this._previousClasses.forEach((name) => {
+          if (!classInfo.includes(name)) {
+            overlay.classList.remove(name);
+            this._previousClasses.delete(name);
+          }
+        });
+      }
+    }
+
+    /**
+     * @param {HTMLElement} overlay
+     * @param {string[]} classInfo
+     * @private
+     */
+    __storeClassNames(overlay, classInfo) {
+      this._previousClasses = new Set();
+
+      this._customClasses = new Set(overlay.className.split(/\s/u).filter((s) => s !== ''));
+
+      classInfo.forEach((name) => {
+        if (!this._customClasses.has(name)) {
+          this._previousClasses.add(name);
+        }
+      });
+    }
+
+    /** @private */
+    __updateOverlayClassNames(overlayClass, overlayElement) {
```

Comment:
Having the 4 different functions that each modify the state it's a bit hard to read. I wonder if all the logic could be included in the main function. Tried the following quick refac and all the tests pass at least:

```js
__updateOverlayClassNames(overlayClass, overlayElement) {
  if (overlayElement) {
    // Overlay is set but overlayClass is not set
    if (overlayClass === undefined) {
      return;
    }
    this.__initialClasses ||= [...overlayElement.classList];

    const classInfo = typeof overlayClass === 'string' ? overlayClass.split(' ') : [];

    if (this._previousClasses !== undefined) {
      // Remove old classes that no longer apply
      this._previousClasses.forEach((name) => {
        if (!this.__initialClasses.includes(name)) {
          overlayElement.classList.remove(name);
        }
      });
    }

    // Add new classes based on the overlayClass
    classInfo.forEach((name) => overlayElement.classList.add(name));

    this._previousClasses = classInfo;
  }
}
```

Reply:
Good point, thanks. I will update the implementation.

Reply:
Updated.

Reply:
>  custom CSS classes can be added / removed also after the first __updateOverlayClassNames call.

The logic that I implemented is mostly aligned with this PR: https://github.com/lit/lit/pull/1764

I think the most important thing is that `overlayClass` can remove the classes it previously set. I don't think the behavior when you have a conflict between a dynamically added class, e.g. set in `renderer` function on the `root`, and `overlayClass` property is important. But should add docs that users shouldn't do that.

Reply:
> Note, custom CSS classes can be added / removed also after the first __updateOverlayClassNames call.

Not sure if we need to take care of this use case since the overlay is only used by `web-components` internally.

Reply:
Updated JSDoc comments and added a note.